src/validation.cpp

```diff
@@ -2794,6 +2795,7 @@ bool PreciousBlock(CValidationState& state, const CChainParams& params, CBlockIn
 bool CChainState::InvalidateBlock(CValidationState& state, const CChainParams& chainparams, CBlockIndex *pindex)
 {
     AssertLockHeld(cs_main);
+    LOCK(::mempool.cs);
```

Comment:
Calling it right before UpdateMempoolForReorg is too fragile and exactly the issue I want to avoid. (E.g. we `DisconnetTip` and then fail to call `UpdateMempoolForReorg`, because another thread acquired the lock)

At the very least the mempool lock should be held as long as disconnectpool sits on the stack.